### PR TITLE
Add support for mixin syntax

### LINKED ISSUE #32: Add support for mixin syntax
Hey,

We recently changed the WebIDL syntax to add support for [mixins][0]. We also added CI tooling to make sure the grammar stayed LL(1), and fixed a couple of issues in the process.

Here's an indicative diff of the recent grammar changes, but please do refer to the [IDL index section][1] in the spec for details:

```diff
--- (old)
+++ (new)
@@ -3,17 +3,13 @@
     ε
 
 Definition ::
-    CallbackOrInterface
+    CallbackOrInterfaceOrMixin
     Namespace
     Partial
     Dictionary
     Enum
     Typedef
-    ImplementsStatement
-
-CallbackOrInterface ::
-    callback CallbackRestOrInterface
-    Interface
+    IncludesStatement
 
 ArgumentNameKeyword ::
     attribute
@@ -23,7 +19,7 @@
     dictionary
     enum
     getter
-    implements
+    includes
     inherit
     interface
     iterable
@@ -38,23 +34,35 @@
     typedef
     unrestricted
 
+CallbackOrInterfaceOrMixin ::
+    callback CallbackRestOrInterface
+    interface InterfaceOrMixin
+
 CallbackRestOrInterface ::
     CallbackRest
-    Interface
+    interface InterfaceRest
+
+InterfaceOrMixin ::
+    InterfaceRest
+    MixinRest
 
-Interface ::
-    interface identifier Inheritance { InterfaceMembers } ;
+InterfaceRest ::
+    identifier Inheritance { InterfaceMembers } ;
 
 Partial ::
     partial PartialDefinition
 
 PartialDefinition ::
-    PartialInterface
+    interface PartialInterfaceOrPartialMixin
     PartialDictionary
     Namespace
 
-PartialInterface ::
-    interface identifier { InterfaceMembers } ;
+PartialInterfaceOrPartialMixin ::
+    PartialInterfaceRest
+    MixinRest
+
+PartialInterfaceRest ::
+    identifier { InterfaceMembers } ;
 
 InterfaceMembers ::
     ExtendedAttributeList InterfaceMember InterfaceMembers
@@ -75,6 +83,22 @@
     : identifier
     ε
 
+MixinRest ::
+    mixin identifier { MixinMembers } ;
+
+MixinMembers ::
+    ExtendedAttributeList MixinMember MixinMembers
+    ε
+
+MixinMember ::
+    Const
+    RegularOperation
+    Stringifier
+    ReadOnly AttributeRest
+
+IncludesStatement ::
+    identifier includes identifier ;
+
 Const ::
     const ConstType identifier = ConstValue ;
 
@@ -130,11 +154,14 @@
     [ ]
 
 Operation ::
-    ReturnType OperationRest
+    RegularOperation
     SpecialOperation
 
+RegularOperation ::
+    ReturnType OperationRest
+
 SpecialOperation ::
-    Special Specials ReturnType OperationRest
+    Special Specials RegularOperation
 
 Specials ::
     Special Specials
@@ -161,8 +188,11 @@
     ε
 
 Argument ::
-    ExtendedAttributeList optional TypeWithExtendedAttributes ArgumentName Default
-    ExtendedAttributeList Type Ellipsis ArgumentName
+    ExtendedAttributeList ArgumentRest
+
+ArgumentRest ::
+    optional TypeWithExtendedAttributes ArgumentName Default
+    Type Ellipsis ArgumentName
 
 ArgumentName ::
     ArgumentNameKeyword
@@ -181,7 +211,7 @@
 
 StringifierRest ::
     ReadOnly AttributeRest
-    ReturnType OperationRest
+    RegularOperation
     ;
 
 StaticMember ::
@@ -189,7 +219,7 @@
 
 StaticMemberRest ::
     ReadOnly AttributeRest
-    ReturnType OperationRest
+    RegularOperation
 
 Iterable ::
     iterable < TypeWithExtendedAttributes OptionalType > ;
@@ -218,7 +248,7 @@
     ε
 
 NamespaceMember ::
-    ReturnType OperationRest
+    RegularOperation
     readonly AttributeRest
 
 Dictionary ::
@@ -229,8 +259,11 @@
     ε
 
 DictionaryMember ::
-    ExtendedAttributeList required TypeWithExtendedAttributes identifier Default ;
-    ExtendedAttributeList Type identifier Default ;
+    ExtendedAttributeList DictionaryMemberRest
+
+DictionaryMemberRest ::
+    required TypeWithExtendedAttributes identifier Default ;
+    Type identifier Default ;
 
 PartialDictionary ::
     dictionary identifier { DictionaryMembers } ;
@@ -259,16 +292,12 @@
 Typedef ::
     typedef TypeWithExtendedAttributes identifier ;
 
-ImplementsStatement ::
-    identifier implements identifier ;
-
 Type ::
     SingleType
     UnionType Null
 
 TypeWithExtendedAttributes ::
-    ExtendedAttributeList SingleType
-    ExtendedAttributeList UnionType Null
+    ExtendedAttributeList Type
 
 SingleType ::
     NonAnyType
```

Let me know if there's anything I can do to help!

[0]: https://github.com/heycam/webidl/commit/45e8173d40ddff8dcf81697326e094bcf8b92920
[1]: https://heycam.github.io/webidl/#idl-grammar

***

*Tracked in: heycam/webidl#472* | *Original pull-request: heycam/webidl#433*

### PR DESCRIPTION
**Pull request by romandev**
This change fixes #32.

### CODE REVIEW

**Comment by romandev** [comment]
@plinss Could you please review this?

File: widlparser/constructs.py [hunk 1]
```diff
@@ -1103,41 +1110,41 @@ def __repr__(self):
         return output + '[argumentlist: ' + (repr(self.arguments) if (self.arguments) else '') + ']]'
 
 
-class ImplementsStatement(Construct):  # [ExtendedAttributes] identifier "implements" identifier ";"
+class IncludesStatement(Construct):  # [ExtendedAttributes] identifier "includes" identifier ";"
```
Comment (TimothyGu) [inline widlparser/constructs.py]: 1. We should retain support for `implements` statements for now, until every spec has been migrated over to `includes`.
2. The production [*IncludesStatement*](https://heycam.github.io/webidl/#prod-IncludesStatement) does not contain a leading `[ExtendedAttributes]`.

File: widlparser/tokenizer.py [hunk 2]
```diff
@@ -54,8 +54,8 @@ class Tokenizer(object):
     SymbolIdents = frozenset((
         'any', 'attribute', 'ArrayBuffer', 'boolean', 'byte', 'ByteString', 'callback', 'const', 'creator', 'DataView',
         'Date', 'deleter', 'dictionary', 'DOMString', 'double', 'enum', 'Error', 'exception', 'false', 'float',
-        'Float32Array', 'Float64Array', 'FrozenArray', 'getter', 'implements', 'Infinity', '-Infinity', 'inherit', 'Int8Array',
-        'Int16Array', 'Int32Array', 'interface', 'iterable', 'legacycaller', 'legacyiterable', 'long', 'maplike',
+        'Float32Array', 'Float64Array', 'FrozenArray', 'getter', 'Infinity', '-Infinity', 'includes', 'inherit', 'Int8Array',
```
Comment (TimothyGu) [inline widlparser/tokenizer.py]: 'includes' < 'Infinity'

File: widlparser/constructs.py [hunk 3]
```diff
@@ -468,6 +469,7 @@ def __init__(self, tokens, parent = None, parser = None):
         Construct.__init__(self, tokens, parent, (not parent), parser = parser)
         self.partial = Symbol(tokens, 'partial') if (Symbol.peek(tokens, 'partial')) else None
         self._interface = Symbol(tokens, 'interface')
+        self.mixin = Symbol(tokens, 'mixin') if (Symbol.peek(tokens, 'mixin')) else None
```
Comment (TimothyGu) [inline widlparser/constructs.py]: There is a difference in the set of children allowed [in an interface](https://heycam.github.io/webidl/#prod-InterfaceMember) versus [in an interface mixin](https://heycam.github.io/webidl/#prod-MixinMember). Specifically, the following are not allowed in interface mixins:

- [Special operations](https://heycam.github.io/webidl/#prod-SpecialOperation) (stringifiers declared on either an operation or an attribute are fine though)
- [Static members](https://heycam.github.io/webidl/#prod-StaticMember)
- [Maplike declarations](https://heycam.github.io/webidl/#prod-ReadWriteMaplike) (readonly or not)
- [Setlike declarations](https://heycam.github.io/webidl/#prod-ReadWriteSetlike) (readonly or not)
- [Inherited attributes](https://heycam.github.io/webidl/#prod-ReadWriteAttribute)

Currently this is not enforced.

**Comment by domenic** [comment]
@romandev @plinss will you have time to carry this to completion?

**Comment by plinss** [comment]
Will take a look later today

**Comment by romandev** [comment]
@plinss, If you don’t mind, can I complete this PR?
I have local WIP patches that I worked on last weekend.
It’s not ready to review yet but I can finish it today.
(In my country, the current time is 8 a.m.)

But if this is urgent, I'm okay even if I don't have a chance to finish this.

**Comment by plinss** [comment]
Sure @romandev, I can defer until tomorrow

**Comment by plinss** [comment]
Looks good with the exception that callbacks can also take mixins. I added that.

**Comment by plinss** [comment]
BTW, many thanks for taking care of this @romandev

**Pull request merged**